node/src/main/scala/org/bitcoins/node/PeerManager.scala

```diff
@@ -1,199 +1,325 @@
-      if (maxConnectedPeers > shuffledPeers.size) {
-        shuffledPeers.take(maxConnectedPeers) ++ getPeersFromDnsSeeds.take(
-          maxConnectedPeers - shuffledPeers.size)
-      } else {
-        shuffledPeers.take(maxConnectedPeers)
+  def awaitPeerWithService(
+      f: ServiceIdentifier => Boolean,
+      timeout: Duration): Future[Unit] = {
+    logger.debug(s"Waiting for peer connection. ${_peerData.keys}")
+    val ret = AsyncUtil
+      .retryUntilSatisfied({
+                             _peerData.exists(x => f(x._2.serviceIdentifier))
+                           },
+                           interval = 1.seconds,
+                           maxTries = timeout.getSeconds.toInt)
+      .recover {
+        case _: AsyncUtil.RpcRetryException =>
+          throw new RuntimeException("No supported peers found!")
+        case unknown: Throwable => throw unknown
       }
-    }
-    allF
+
+    ret
   }
 
-  def addPeer(peer: Peer): Unit = {
-    if (!_peerData.contains(peer))
-      _peerData.put(peer, PeerData(peer, node))
-    else logger.debug(s"Peer $peer already added.")
-    ()
+  def replacePeer(replacePeer: Peer, withPeer: Peer): Future[Unit] = {
+    logger.debug(s"Replacing $replacePeer with $withPeer")
+    assert(!peerData(replacePeer).serviceIdentifier.nodeCompactFilters,
+           s"$replacePeer has cf")
+    removePeer(replacePeer)
+    addPeer(withPeer)
   }
 
-  def removePeer(peer: Peer): Future[Unit] = {
-    if (_peerData.contains(peer)) {
-      val connF = peerData(peer).peerMessageSender.isConnected()
-      val disconnectF = connF.map { conn =>
-        if (conn) peerData(peer).peerMessageSender.disconnect()
-        else Future.unit
+  def removePeer(peer: Peer): Unit = {
+    logger.debug(s"Removing persistent peer $peer")
+    val client = peerData(peer).client
+    _peerData.remove(peer)
+    //so we need to remove if from the map for connected peers so no more request could be sent to it but we before
+    //the actor is stopped we don't delete it to ensure that no such case where peers is deleted but actor not stopped
+    //leading to a memory leak may happen
+    _waitingForDeletion.add(peer)
+    //now send request to stop actor which will be completed some time in future
+    client.close()
+  }
+
+  def isReconnection(peer: Peer): Boolean = {
+    peerData.contains(peer)
+  }
+
+  def start: Future[Unit] = {
+    logger.debug(s"Starting PeerManager")
+    finder.start.map(_ => logger.info("Done starting PeerManager"))
+  }
+
+  def peerData: Map[Peer, PeerData] = _peerData.toMap
+
+  def stop: Future[Unit] = {
+    logger.debug(s"Stopping PeerManager")
+
+    val finderStopF = finder.stop
+
+    peers.foreach(removePeer)
+
+    val managerStopF = AsyncUtil.retryUntilSatisfied(
+      _peerData.isEmpty && waitingForDeletion.isEmpty,
+      interval = 1.seconds,
+      maxTries = 10)
+
+    for {
+      _ <- finderStopF
+      _ <- managerStopF
+    } yield ()
```

Comment:
another log indicating that it fully stopped 
```scala
   for {
      _ <- finderStopF
      _ <- managerStopF
    } yield {
 logger.info(s"Stopped peer manager")
}
```
a timer might be useful here to to see how long it took to stop